modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/components/criteria_builder/CriteriaGroup.es.js

```diff
@@ -0,0 +1,368 @@
+ */
+const withDragSource = dragSource(
+	DragTypes.CRITERIA_GROUP,
+	{
+		beginDrag
+	},
+	(connect, monitor) => ({
+		connectDragPreview: connect.dragPreview(),
+		connectDragSource: connect.dragSource(),
+		dragging: monitor.isDragging()
+	})
+);
+
+class CriteriaGroup extends Component {
+	static propTypes = {
+		connectDragPreview: PropTypes.func,
+		connectDragSource: PropTypes.func,
+		criteria: PropTypes.object,
+		dragging: PropTypes.bool,
+		editing: PropTypes.bool,
+		groupId: PropTypes.string,
+		index: PropTypes.number,
+		modelLabel: PropTypes.string,
+		onChange: PropTypes.func,
+		onMove: PropTypes.func,
+		parentGroupId: PropTypes.string,
+		root: PropTypes.bool,
+		supportedConjunctions: PropTypes.array,
+		supportedOperators: PropTypes.array,
+		supportedProperties: PropTypes.array,
+		supportedPropertyTypes: PropTypes.object
+	};
+
+	static defaultProps = {
+		root: false
+	};
+
+	constructor(props) {
+		super(props);
+
+		this.NestedCriteriaGroupWithDrag = withDragSource(CriteriaGroup);
+	}
+
+	_getConjunctionLabel(conjunctionName, conjunctions) {
+		const conjunction = conjunctions.find(
+			({name}) => name === conjunctionName
+		);
+
+		return conjunction ? conjunction.label : undefined;
+	}
+
+	_handleConjunctionClick = event => {
+		event.preventDefault();
+
+		const {criteria, onChange, supportedConjunctions} = this.props;
+
+		const index = supportedConjunctions.findIndex(
+			item => item.name === criteria.conjunctionName
+		);
+
+		const conjunctionSelected = index === supportedConjunctions.length - 1 ?
+			supportedConjunctions[0].name :
+			supportedConjunctions[index + 1].name;
+
+		onChange(
+			{
+				...criteria,
+				...{
+					conjunctionName: conjunctionSelected
+				}
+			}
+		);
+	}
+
+	/**
+	 * Adds a new criterion in a group at the specified index. If the criteria
+	 * was previously empty and is being added to the root group, a new group
+	 * will be added as well.
+	 * @param {number} index The position the criterion will be inserted in.
+	 * @param {object} criterion The criterion that will be added.
+	 * @memberof CriteriaGroup
+	 */
+	_handleCriterionAdd = (index, criterion) => {
+		const {criteria, onChange, root, supportedOperators} = this.props;
+
+		const {operatorName, propertyName, value} = criterion;
+
+		const newCriterion = {
+			operatorName: operatorName ?
+				operatorName :
+				supportedOperators[0].name,
+			propertyName,
+			value: value ? value : ''
+		};
+
+		if (root && !criteria) {
+			onChange(
+				{
+					conjunctionName: CONJUNCTIONS.AND,
+					groupId: generateGroupId(),
+					items: [newCriterion]
+				}
+			);
+		}
+		else {
+			onChange(
+				{
+					...criteria,
+					...{
+						items: insertAtIndex(
+							newCriterion,
+							criteria.items,
+							index
+						)
+					}
+				}
+			);
+		}
+	}
+
+	_handleCriterionChange = index => newCriterion => {
+		const {criteria, onChange} = this.props;
+
+		onChange(
+			{
+				...criteria,
+				...{
```

Comment:
^

btw, I usually just make this note so that its easier for you to track. Not trying to be obnoxious lol